apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -263,43 +263,49 @@ public void beforeEnd(long epochMicros) {
     @Override
     protected void afterEnd() {
         if (transaction != null && transaction.isSpanCompressionEnabled() && parent != null) {
-            Span buffered = parent.bufferedSpan.get();
-            if (parent.isFinished() || !isCompressionEligible()) {
-                if (buffered != null) {
-                    if (parent.bufferedSpan.compareAndSet(buffered, null)) {
-                        this.tracer.endSpan(buffered);
+            Span parentBuffered = parent.bufferedSpan.incrementReferencesAndGet();
+            try {
+                if (parent.isFinished() || !isCompressionEligible()) {
+                    if (parentBuffered != null) {
+                        if (parent.bufferedSpan.compareAndSet(parentBuffered, null)) {
+                            this.tracer.endSpan(parentBuffered);
+                        }
                     }
-                }
-                parent.decrementReferences();
-                this.tracer.endSpan(this);
-                return;
-            }
-            if (buffered == null) {
-                parent.decrementReferences();
-                if (!parent.bufferedSpan.compareAndSet(null, this)) {
-                    // the failed update would ideally lead to a compression attempt with the new buffer,
-                    // but we're dropping the compression attempt to keep things simple
+                    parent.decrementReferences();
                     this.tracer.endSpan(this);
+                    return;
                 }
-                return;
-            }
-            if (!buffered.tryToCompress(this)) {
-                if (parent.bufferedSpan.compareAndSet(buffered, this)) {
+                if (parentBuffered == null) {
                     parent.decrementReferences();
-                    this.tracer.endSpan(buffered);
-                } else {
+                    if (!parent.bufferedSpan.compareAndSet(null, this)) {
+                        // the failed update would ideally lead to a compression attempt with the new buffer,
+                        // but we're dropping the compression attempt to keep things simple
+                        this.tracer.endSpan(this);
+                    }
+                    return;
+                }
+                if (!parentBuffered.tryToCompress(this)) {
+                    if (parent.bufferedSpan.compareAndSet(parentBuffered, this)) {
+                        parent.decrementReferences();
+                        this.tracer.endSpan(parentBuffered);
+                    } else {
+                        parent.decrementReferences();
+                        // the failed update would ideally lead to a compression attempt with the new buffer,
+                        // but we're dropping the compression attempt to keep things simple
+                        this.tracer.endSpan(this);
+                    }
+                } else if (isSampled()) {
```

Comment:
We never decrement the `parent` references if `isSampled()==false` here. It was the same prior to your change, but is this intentional?